RFC: Dust Protection (#32)
iotaledger/protocol-rfcs: text/0032-dust-protection/0032-dust-protection.md

+ Feature name: dust-protection
+ Start date: 2020-12-07
+ RFC PR: [iotaledger/protocol-rfcs#0032](https://github.com/iotaledger/protocol-rfcs/pull/0032)

# Summary

> 

Reviewer: ```suggestion
In the UTXO model, each node in the network needs to keep track of all the currently unspent outputs. When the number of outputs gets too large, this can cause performance and memory issues.
```
Maybe add one introduction sentence

> This RFC defines a new protocol rule regarding processing outputs that transfer a small amount of Iotas, known as dust outputs. The rule states that each dust output should be backed up by a deposit of the receiving address. Thus making it expensive to proliferate dust. Since a receiver must make a deposit, the protocol makes receiving dust an opt-in feature.

Reviewer: ```suggestion
This RFC proposes a new protocol rule regarding the processing of outputs that transfer a very small amount of IOTA, so-called dust outputs: Dust outputs are only allowed when they are backed up by a certain deposit on the receiving address. This limits the amount of dust outputs Thus making it expensive to proliferate dust. Since a receiver must make a deposit, the protocol makes receiving dust an opt-in feature.
```

# Motivation

> The total supply of IOTA is approximately 2.78 peta IOTA. Considering that a single typical `SigLockedSingleOutput` takes 42 bytes (just the plain output bytes without the encompassing transaction), it is easy to see how the size of the UTXO set may pose a potential problem. An attacker splitting 1 GI to single iota outputs can blow up the set size so that it won't fit into 4GB of RAM. This is far from being a prohibitively expensive attack.

Reviewer: ```suggestion
A single typical `SigLockedSingleOutput` as described in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md) has a size of 42 bytes. When 1 Gi is split up into one billion outputs of 1 IOTA each, this corresponds to about 4 GB of outputs that need to be processed and stored. Since IOTA is feeless, this is far from a prohibitively expensive attack.
```
Reference the transaction RFC, it's not primarily an issue of RAM but more DB storage and processing in general.

Author: Not really sure it is not  issue of RAM primarily...
Sure you can always do I/O swaps so the node won't crash...

It will slow down the node considerably though..

> Thus, in order to protect nodes from such memory attacks we should make accumulating small dust outputs expensive. Other DLTs have fees that provide protection, but IOTA is feeless. So instead of fees we can have deposits. 

Reviewer: ```suggestion
In order to protect nodes from such attacks, one possible solution is to make accumulating dust outputs expensive. Since IOTA does not have any fees that might limit the feasibility of issuing many dust transactions, deposits pose a valid alternative to achieve a similar effect.
```

> When an address wants to receive micro transactions it must have a special output locked as a deposit. This deposit can't be consumed by any transaction as long as the dust outputs exist on the address. This also prevents attackers from sending dust to addresses that didn't opt into allowing dust outputs.

Reviewer 2: This partially conflicts with what stated later: that one can consume the Deposit UTXOs as long as the remaining deposit(s) cover the unspent dust UTXOs.

> When an address wants to receive micro transactions it must have a special output locked as a deposit. This deposit can't be consumed by any transaction as long as the dust outputs exist on the address. This also prevents attackers from sending dust to addresses that didn't opt into allowing dust outputs.

Reviewer: ```suggestion
When an address is supposed to receive micro transactions, it must have an unspent output of a special type as a deposit. This deposit cannot be consumed by any transaction as long as the dust outputs remain unspent.
```

> Additional benefit of this rule is that it makes a mass of privacy violating [forced address reuse attacks](https://en.bitcoin.it/wiki/Privacy#Forced_address_reuse) more expensive to carry out.

Reviewer: ```suggestion
An additional benefit of this rule is that it makes a mass of privacy violating [forced address reuse attacks](https://en.bitcoin.it/wiki/Privacy#Forced_address_reuse) more expensive to carry out.
```

# Detailed design

### Definitions

> *Dust*: a UTXO that spends an amount smaller than 1 MI.

Reviewer: ```suggestion
*Dust output*: A transaction output that has an amount smaller than 1 Mi
```
For the general definition Dust outputs don't necessarily need to be unspent

Author: btw, when you go to exchanges and crypto listings (at least the ones I checked out), they usually write `MI`

> *SigLockedDustAllowanceOutput* - A new output type for deposits that allow an address to recieve dust outputs. It can be consumed as an input like a regular SigLockedOutput.

Reviewer 2: Who can generate this output for the address? An address that does not want to support dust could be forced into it.

Reviewer: Yes, a 3rd party can create a deposit on your address. But it this really an issue? You have the option to just ignore any dust and SigLockedDustAllowanceOutput and only consider your expected transactions. This would then effectively be equivalent to forbidding foreign SigLockedDustAllowanceOutput. Or, you even have the option to siphon the dust and make (small) profit. 

Forbidding foreign SigLockedDustAllowanceOutput would be possible but also complicate the validation and introduce some strange criteria here. I think, we talked about this, but keeping everything as simple as possible seemed more desirable especially since it is not a big drawback.

However, the least we should do is mention this and our motivation.

> *SigLockedDustAllowanceOutput* - A new output type for deposits that allow an address to recieve dust outputs. It can be consumed as an input like a regular SigLockedOutput.

Reviewer: ```suggestion
*SigLockedDustAllowanceOutput*: A new output type for deposits that enables an address to receive dust outputs. It can be consumed as an input like a regular `SigLockedSingleOutputs`.
```

> <table>
>                                     <tr>
>                                         <td><b>Name</b></td>
>                                         <td><b>Type</b></td>
>                                         <td><b>Description</b></td>
>                                     </tr>
>                                     <tr>
>                                         <td>Output Type</td>
>                                         <td>uint8</td>
>                                         <td>
>                                             Set to <strong>value 1</strong> to denote a <i>SigLockedDustAllowanceOutput</i>.
>                                         </td>
>                                     </tr>
>                                     <tr>
>                                         <td valign="top">Address </td>
>                                         <td colspan="2">
>                                                                      <details>
>                                                 <summary>Ed25519 Address</summary>
>                                                 <table>
>                                                     <tr>
>                                                         <td><b>Name</b></td>
>                                                         <td><b>Type</b></td>
>                                                         <td><b>Description</b></td>
>                                                     </tr>
>                                                     <tr>
>                                                         <td>Address Type</td>
>                                                         <td>uint8</td>
>                                                         <td>
>                                                             Set to <strong>value 1</strong> to denote an <i>Ed25519 Address</i>.
>                                                         </td>
>                                                     </tr>
>                                                     <tr>
>                                                         <td>Address</td>
>                                                         <td>Array&lt;byte&gt;[32]</td>
>                                                         <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
>                                                     </tr>
>                                                 </table>
>                                             </details>
>                                         </td>
>                                     </tr>
>                                     <tr>
>                                         <td>Amount</td>
>                                         <td>uint64</td>
>                                         <td>The amount of tokens to deposit with this <i>SigLockedSingleOutput</i> output.</td>
>                                     </tr>
>                                 </table>

Reviewer: ```suggestion
<table>
  <tr>
    <td><b>Name</b></td>
    <td><b>Type</b></td>
    <td><b>Description</b></td>
  </tr>
  <tr>
    <td>Output Type</td>
    <td>uint8</td>
    <td>
      Set to <strong>value 1</strong> to denote a <i>SigLockedDustAllowanceOutput</i>.
    </td>
  </tr>
  <tr>
    <td valign="top">Address</td>
    <td colspan="2">
      <details>
        <summary>Ed25519 Address</summary>
        <table>
          <tr>
            <td><b>Name</b></td>
            <td><b>Type</b></td>
            <td><b>Description</b></td>
          </tr>
          <tr>
            <td>Address Type</td>
            <td>uint8</td>
            <td>
              Set to <strong>value 1</strong> to denote an <i>Ed25519 Address</i>.
            </td>
          </tr>
          <tr>
            <td>Address</td>
            <td>Array&lt;byte&gt;[32]</td>
            <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Amount</td>
    <td>uint64</td>
    <td>The amount of tokens to deposit with this <i>SigLockedSingleOutput</i> output.</td>
  </tr>
</table>
```
Beautify formatting

> ### The new protocol dust rule

Reviewer 2: I think this validation rules should be rewritten more technically, even in pseudo-code; as there are some ambiguities on the order of processing, especially in the context of WhiteFlag.

Reviewer: I am not even sure, whether we need explain this in the context of WhiteFlag. The transaction RFC only talks about "Semantic validation" of transactions, and semantically invalid transactions are handled via white flag. So I think it would be best to just describe the dust validation in the same way splitting up into Syntactical Validation and Semantic Validation and then refer to the transaction RFC

Author: Pseudo code added

> ### The new protocol dust rule

Reviewer: ```suggestion
### Validation
```
Let's call this validation

Let `A` be the address that should hold the dust outputs' balances. Let `M` be the sum of all SigLockedDustAllowanceOutputs' values on `A`.So the total allowed dust outputs on `A` is (`M`/1 MIOTA) * 100 rounded down to the nearest integer. So 100 outputs for 1 MIOTA deposited.

The SigLockedDustAllowanceOutput must be greater or equal to 1MI.

> If while processing a milestone a new dust UTXO is created but there is not a sufficient amount already locked in SigLockedDustAllowanceOutputs then the encompassing message should be marked as `ignored`.

Reviewer 2: I believe `ignored` here is a reference to the whiteflag RFC, put link reference.

> If while processing a milestone a new dust UTXO is created but there is not a sufficient amount already locked in SigLockedDustAllowanceOutputs then the encompassing message should be marked as `ignored`.

Reviewer 2: This introduces a race condition: an attacker could always create a new dust UTXO between milestones epochs for the unlocking of the deposit UTXO to be ignored. While this will perhaps not ever be a money issue (1MIOTA is a negligible amount of money) it may represent a processing issue: some use-case may rely on this deposit and address to be invalidated for good. Therefore, I think we need to define better how dust consolidation and deposit unlock look like.

Author: In our discussions we considered this not a serious issue

We assumed that the receiver can always consolidate the attacker's outputs and just get free money as long as the attack continues. However, your point have to be validated with bizdev probably

Reviewer: Yes, there is some race-condition/censoring where an attacker always tries to issue a dust transaction before (wrt to the WhiteFlag order) the spend transaction of a `SigLockedDustAllowanceOutputs`. I guess this is exactly, what you guys are talking about?
I think this is an important issue and we should talk about it in the RFC. This has been one of my main concerns with the dust protection without a special output type. But I think it is much less important in this scenario: First, you can just ignore SigLockedDustAllowanceOutputs, and just move everything else without any possibilities for race-conditions (this should be sufficient for the usual "wallet" use-case). Secondly, when you are under attack and really need to spend everything, you can still "dust yourself" until the limit is reached and then issue one big transaction consuming everything at once.

Reviewer 2: > First, you can just ignore SigLockedDustAllowanceOutputs, and just move everything else without any possibilities for race-conditions (this should be sufficient for the usual "wallet" use-case).

I don't get this first point. Are you suggesting a special WF "ordering" of dust-related transactions?

Reviewer: I mean, as an honest user whose address got polluted by an attacker sending a SigLockedDustAllowanceOutputs and tons of dust outputs, you can just ignore all the dust and only transfer away regular SigLockedOutputs of >= 1 Mi. When doing this there cannot be any race condition and it would be the exact same result as when an honest users could "block" SigLockedDustAllowanceOutputs.

This is true without any changes to the validation or WF ordering.

Reviewer 2: That scenario is clear: as an attacker, you cannot deny a user to spend their regular funds. What I was referring to is that, as an attacker, you can deny a user to spend its last `SigLockedDustAllowanceOutput`.

Author: This md discusses the issue at length and offers mitigations:
https://hackmd.io/@iota-protocol/S1zFaBCjP

Reviewer 3: I told wollac yesterday about a special "lock type" to stop collecting dust. This lock type needs to be signed with the private key for the address. This way you could stop collecting dust in the first milestone, and spend your funds without a race condition in the next milestone.
We came to the conclusion that we won't add this lock type now, but I just wanted to write it down, so the idea is not lost.

Author: Why did you come to the conclusion that this shouldn't be added?
Doesn't sound too complicated

> Let `A` be the address that should hold the dust outputs' balances. Let `M` be the sum of all SigLockedDustAllowanceOutputs' values on `A`.So the total allowed dust outputs on `A` is (`M`/1 MIOTA) * 100 rounded down to the nearest integer. So 100 outputs for 1 MIOTA deposited.
> 
> The SigLockedDustAllowanceOutput must be greater or equal to 1MI.
> 
> If while processing a milestone a new dust UTXO is created but there is not a sufficient amount already locked in SigLockedDustAllowanceOutputs then the encompassing message should be marked as `ignored`.
> 
> If while processing a milestone an unspent SigLockedDustAllowanceOutput is consumed and as a result we have a violation on the number of allowed dust outputs then the encompassing message should be marked as `ignored`.
> 
> The checks should happen after the entire message was processed. Messages that were previously applied by the current milestone are taken into account.

Reviewer: ```suggestion
Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

The amount of a `SigLockedDustAllowanceOutput` must be at least 1 Mi. Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedOutputs`. The transaction validation as defined in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md), however, needs to be adapted.

_Syntactical validation_ for `SigLockedDustAllowanceOutput`:
- The `Address` must be unique in the set of `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`.
- The `Amount` must be ≥ 1,000,000.

The _semantic validation_ remains unchanged and are checked for both `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`, but this RFC introduces one additional criterion:

A transaction T
  - consuming a `SigLockedDustAllowanceOutput` on address A **or**
  - creating a dust output with address A,

is only semantically valid, if, after T is booked, the number of unspent dust outputs on A must not exceed the allowed threshold of S / 10,000. When T contains more than one such output, the threshold must not be exceeded on any corresponding address.
```
Made the validation rules more technical and consistent with the way it is described in the transaction RFC

Reviewer 2: > Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedDustAllowanceOutput`. 

?

Reviewer 2: > - The `Address` must be unique in the set of `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`.

This does not need to be true as long as each output of type `SigLockedDustAllowanceOutput` has an amount of >= 1,000,000 .

Reviewer 2: is only semantically valid, if, after T is booked, the number of unspent dust outputs on A ~~must~~ **does** not exceed the allowed threshold of S / 10,000.

---

When T contains more than one such ~~output~~ input / consumes more than one such output

Reviewer: > You mean `TransactionPayload`?

What I mean is that in #18 any mention of `SigLockedSingleOutput` is supposed to be replaced with "`SigLockedSingleOutput` and/or `SigLockedDustAllowanceOutput`". But I didn't wanna go through all occurrences explicitly. Any Ideas how to makes this more clear.

Reviewer: > This does not need to be true as long as each output of type `SigLockedDustAllowanceOutput` has an amount of >= 1,000,000 .

Currently #18 forbids creating a transaction with more than one output to the same address. We should try to preserve this somehow. But indeed since we now have more than one output type the question is, do we want to allow a transaction with 1 `SigLockedDustAllowanceOutput` to A and 1 `SigLockedSingleOutput` to A? Or is this forbidden?

Definitely we should forbid more than one  `SigLockedDustAllowanceOutput` to A in one transaction, just we currently do for `SigLockedSingleOutput`.

Reviewer: > When T contains more than one such ~output~ input / consumes more than one such output

Not 100% sure whether I understand this remark. I made some minimal changes to the previously suggested changes, is that better now?

Author: Not sure I understand that last sentence you added...
at least I can interpret it more than 1 way

Reviewer: OK then. I've removed that sentence, it is redundant anyways, I was just trying to make it more explicit that when a transaction contains multiple dust outputs on different addresses, the criteria must hold for all of them.

# Drawbacks

> 1. We can't have an address that will ever hold less than 1 MI. It will be a bit hard on testnet faucets but still workable.

Reviewer: ```suggestion
- There can no longer be addresses holding less than 1 Mi.
```
I don't think testnet faucets would be the main issue here, this could matter for quite a few projects, so maybe just leave this example out.

> 2. A service receiving micropayments may fail receiving them if it didn't consolidate dust outputs or raised the deposit for the address.

Reviewer: ```suggestion
- A service receiving micropayments may fail receiving them, if it did not consolidate dust outputs or raised the deposit for the receiving address.
```

# Rationale and alternatives

> Other solutions that were discussed:

Reviewer: ```suggestion
There are potential alternatives to introducing dust deposits:
```

> *Burning Dust* - allow dust to live for a limited amount of time on the ledger before removing it from total supply.

Reviewer: ```suggestion
- *Burning dust*: Allow dust outputs to exists only for a limited amount of time in the ledger before removing them.
```

Reviewer 2: mentioned of the altered total supply needs to be introduced, as it is not clear from your suggestion were these burned UTXOs go to.

Reviewer: Tried to clarify this in the suggested change

> 
> *Sweeping dust into merkle tree* - Instead of storing all of the dust in the UTXO set, some nodes can accumulate the dust outputs to a merkle tree root. When the dust is spent, those nodes will require a merkle proof of inclusion.

Reviewer: ```suggestion
- *Sweeping dust into Merkle trees*: Instead of burning dust outputs after some time, they are instead compressed into a Merkle tree and only the tree root is kept. In order to spend one of these compressed outputs, the corresponding Merkle audit path needs to be supplied in addition to a regular signature.
```

I think the consideration which nodes do this sweeping or not is a bit confusing and very closely related to any permanode issues we might currently have.

> The first solution has some attack vectors that may burn dust that a user may have wanted to consolidate. Also solutions changing the supply can be quite controversial.

Reviewer: ```suggestion
The first option can cause issues, when dust outputs were burned before users could consolidate them. Also, changing the supply can be controversial.
```

> The second solution, besides being more complicated, will relief the problem from one class of nodes but not all nodes.

Reviewer: ```suggestion
The second option is much more complicated as it introduces a completely new unlock mechanisms and requires the nodes to store the Merkle tree roots indefinitely.
```

Reviewer 2: I think user-side added complexity should also be mentioned: a user need a way to be notified about the compression of his dust outputs into a certain merkle root, and need a way to efficiently query some permanode to obtain the corresponding audit path, bla bla. No need to be super-specific but just underline the fact that it exposes tons of complexities.

# Unresolved questions

> - An attacker may try to fill an address with dust to fill the total cap allowed and block honest senders that want to pay the service. The service can protect by simply consolidating the attacker dust and collecting it for profit. The problem is that if the cost of doing PoW too often may exceed the bad dust profit. Perhaps it is a good idea to limit "tiny" (1 i) dust outputs more than larger dust outputs? In the original discussion each dust class was supposed to be capped to a certain fixed amount. Perhaps we should give weights to different dust classes? So "tiny" dust will weigh more and fill the cap more quickly than larger dust?

Reviewer: ```suggestion
- An attacker can send microtransactions to an address with a `SigLockedDustAllowanceOutput` in order to fill the allowed threshold and block honest senders of microtransactions. The owner of the address can mitigate this by simply consolidating the attacker's dust and collecting it for profit. The problem is that if the cost of doing PoW too often may exceed the bad dust profit. Perhaps it is a good idea to limit "tiny" (1 i) dust outputs more than larger dust outputs? In the original discussion each dust class was supposed to be capped to a certain fixed amount. But perhaps we should give weights to different dust classes. So "tiny" dust will weigh more and fill the cap more quickly than larger dust?
```

Reviewer 2: The problem is that ~~if~~ the cost of doing PoW too often may exceed the ~~bad dust~~ profit made by collecting dust.

Reviewer: integrated in my suggested change

- Total cap per address so db key won't be polluted... should this be part of the RFC? This also depends on how key-value dbs will behave at the implementation level. RocksDb for example can configured to handle this well.

> - If we stay with the naive linear scheme, is 100 outputs per 1MI is really our choice?

Reviewer 2: This is a tricky point, since the deposit/dust_outputs ratio threshold can only be decreased once in place, otherwise we'd invalidate confirmed UTXOs. So we should rather start with a really conservative (high) deposit for dust UTXOs number.

Reviewer: Is this really an issue? Since the validity of dust transaction is only evaluated when they become confirmed, it would still work as expected even though the dust state on certain address might technically be invalid.

Reviewer 2: It is true that you are not invalidating the unlock condition, but just the confirmation condition; nonetheless it will undermine validation of the past, in which case you will have to strictly rely on proofs of inclusions, without being able to make up your own mind on what is valid or not.

Author: > This is a tricky point, since the deposit/dust_outputs ratio threshold can only be decreased once in place
The ratio can't be decreased below the validation constraints

        <table>
          <tr>
            <td><b>Name</b></td>
            <td><b>Type</b></td>
            <td><b>Description</b></td>
          </tr>
          <tr>
            <td>Address Type</td>
            <td>uint8</td>
            <td>
              Set to <strong>value 1</strong> to denote an <i>Ed25519 Address</i>.
            </td>
          </tr>
          <tr>
            <td>Address</td>
            <td>Array&lt;byte&gt;[32]</td>
            <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Amount</td>
    <td>uint64</td>
    <td>The amount of tokens to deposit with this <i>SigLockedDustAllowanceOutput</i> output.</td>
  </tr>
</table>

### Validation

Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 100,000 and rounded down, i.e. 10 outputs for each 1 Mi deposited.
However, regardless of S, the number of dust outputs must never exceed 100 per address.

> The amount of a `SigLockedDustAllowanceOutput` must be at least 1 Mi. Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedSingleOutput`. The transaction validation as defined in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md), however, needs to be adapted.

Reviewer 4: > The amount of a `SigLockedDustAllowanceOutput` must be at least 1 Mi.

This is already said right after in the syntactical validation, bit repetitive.

+ Feature name: dust-protection
+ Start date: 2020-12-07
+ RFC PR: [iotaledger/protocol-rfcs#0032](https://github.com/iotaledger/protocol-rfcs/pull/0032)

# Summary
In the UTXO model, each node in the network needs to keep track of all the currently unspent outputs. When the number of outputs gets too large, this can cause performance and memory issues.
This RFC proposes a new protocol rule regarding the processing of outputs that transfer a very small amount of IOTA, so-called dust outputs: Dust outputs are only allowed when they are backed up by a certain deposit on the receiving address. This limits the amount of dust outputs, thus making it expensive to proliferate dust. Since a receiver must make a deposit, the protocol makes receiving dust an opt-in feature.

# Motivation

> A single typical `SigLockedSingleOutput` as described in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md) has a size of 42 bytes. When 1 Gi is split up into one billion outputs of 1 IOTA each, this corresponds to about 4 GB of outputs that need to be processed and stored. Since IOTA is feeless, this is far from a prohibitively expensive attack.

Reviewer: ```suggestion
A single typical `SigLockedSingleOutput` as described in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md) has a size of 42 bytes. When 1 Gi is split up into one billion outputs of 1 IOTA each, this corresponds to about 40 GB of outputs that need to be processed and stored. Since IOTA is feeless, this is far from a prohibitively expensive attack.
```
1 billion * 42 bytes ≈ 40GB

Author: yikes, missing 0

  <tr>
    <td><b>Name</b></td>
    <td><b>Type</b></td>
    <td><b>Description</b></td>
  </tr>
  <tr>
    <td>Output Type</td>
    <td>uint8</td>
    <td>
      Set to <strong>value 1</strong> to denote a <i>SigLockedDustAllowanceOutput</i>.
    </td>
  </tr>
  <tr>
    <td valign="top">Address</td>
    <td colspan="2">
      <details>
        <summary>Ed25519 Address</summary>
        <table>
          <tr>
            <td><b>Name</b></td>
            <td><b>Type</b></td>
            <td><b>Description</b></td>
          </tr>
          <tr>
            <td>Address Type</td>
            <td>uint8</td>
            <td>
              Set to <strong>value 1</strong> to denote an <i>Ed25519 Address</i>.
            </td>
          </tr>
          <tr>
            <td>Address</td>
            <td>Array&lt;byte&gt;[32]</td>
            <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Amount</td>
    <td>uint64</td>
    <td>The amount of tokens to deposit with this <i>SigLockedDustAllowanceOutput</i> output.</td>
  </tr>
</table>

### Validation

> Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

Reviewer 3: We should add an optional query paramter to the API calls to decide if we want to get all UTXO on an address, or only all non-dust UTXO, or only all dust UTXO (to sweep them in the client).

This is needed for the following API calls:

```go
	// RouteAddressBech32Outputs is the route for getting all output IDs for an address.
	// The address must be encoded in bech32.
	// GET returns the outputIDs for all outputs of this address (optional query parameters: "include-spent").
	RouteAddressBech32Outputs = "/addresses/:address/outputs"

	// RouteAddressEd25519Outputs is the route for getting all output IDs for an ed25519 address.
	// The ed25519 address must be encoded in hex.
	// GET returns the outputIDs for all outputs of this address (optional query parameters: "include-spent").
	RouteAddressEd25519Outputs = "/addresses/ed25519/:address/outputs"
```

> Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

Reviewer 3: This is kinda off-topic, and only my personal opinion, but we should add another special output type if we want to unlock an address for a colored coin.
So we don't mix the dust protection logic with the colored coin logic.
For example I may want to use colored coins as a user, but I don't want an attacker to be able to dust my address.

> Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

Author: - I guess there's a bigger threat from users misusing the protocol than actual attackers. It will just mean that users will have to create the dust allowance over several messages which is not so bad.
- Since dust collecting addresses should be encouraged to sweep as soon as possible I think putting this limit is a good idea.
- Definitely we should add those queries. Was waiting for that meeting with Charlie before requesting changes for the API RFC
- Maybe, not sure what the spec for colored coins will look like yet :sweat_smile: 

@Wollac, @luca-moser , @thibault-martinez , and @karimodm 
give :+1: if you agree with those modifications

> Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

Reviewer: - I am all for a total upper limit, that was still an open question in the RFC. However, I've no idea whether 1000 is the right parameter here.
- I am not so sure about counting all outputs the same, no matter what there actual value is. With an upper limit, the Ti accident has already been resolved, hasn't it? To me this is just needlessly complicates the creation of dust addresses, doesn't provide any attack protection and is not really easier to implement on the node-side.

> Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

Reviewer 3: You are right, this way it is even better.
So leave it balance based and introduce an upper limit. :+1:

          <tr>
            <td>Address</td>
            <td>Array&lt;byte&gt;[32]</td>
            <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Amount</td>
    <td>uint64</td>
    <td>The amount of tokens to deposit with this <i>SigLockedDustAllowanceOutput</i> output.</td>
  </tr>
</table>

### Validation

Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.

The amount of a `SigLockedDustAllowanceOutput` must be at least 1 Mi. Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedSingleOutput`. The transaction validation as defined in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md), however, needs to be adapted.

_Syntactical validation_ for `SigLockedDustAllowanceOutput`:

> - The `Address` must be unique in the set of `SigLockedDustAllowanceOutputs` in one transaction T. However, there can be one `SigLockedSingleOutput` and one `SigLockedDustAllowanceOutputs` T.

Reviewer 5: What's the motivation behind allowing a `SigLockedSingleOuput` + `SigLockedDustAllowanceOutput` within the same transaction targeting the same address?

Author: I think it makes more sense that the validation rules will be enforced per output type.

It can also be useful if someone wishes to reuse and address. Imagine someone having funds in an address. Now we want to split the funds so part of them will be the allowance without moving anything out of the address.

What would be the benefits of changing this rule?

- The `Amount` must be ≥ 1,000,000.

The _semantic validation_ remains unchanged and are checked for both `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`, but this RFC introduces one additional criterion:

A transaction T
  - consuming a `SigLockedDustAllowanceOutput` on address A **or**
  - creating a dust output with address A,

is only semantically valid, if, after T is booked, the number of unspent dust outputs on A does not exceed the allowed threshold of S / 10,000.

The following pseudocode demonstrates the new rules:

```python
boolean DustValidation(Transaction tx):
    # maps return 0 for non-existent keys
    dustAllowanceAddressesToBalances = new map()
    dustAddressesToNumOfDustOutputs = new map()
    
    if tx not null:
        for output in tx.outputs:
            if output.type == SigLockedDustAllowanceOutput:
                # must be at least 1 Mi
                if output.balance < 1_000_000:
                    return false
                # only one allowance deposit per address per transaction

>                 if (dustAllowanceAddressesToBalances[output.address] == 0):

Reviewer 2: Here you are treating the map as initially empty...

                    dustAllowanceAddressesToBalances[output.address] = output.amount
                else
                    return false
            # count new dust outputs
            if output.type == SigLockedSingleOutput & output.balance < 1_000_000:
                dustAddressesToNumOfDustOutputs[output.address] += 1
        for input in tx.inputs:
            # load from the UTXO set
            output = UTXO.loadUtxoForInput(input)
            if output.amount < 1_000_000:
               dustAddressesToNumOfDustOutputs[output.address] -= 1
            if output.type == SigLockedDustAllowanceOutput:

>                 dustAllowanceAddressesToBalances[output.address] =- output.amount

Reviewer 2: ...but here you assume that some initial deposit could be there, consumed in the input. You may therefore ending up with a value mapped for that deposit < 0. Take the case that you have a Tx that has an input SigLockedDustAllowanceOutput UTXO of 1.1Mi and a deposit output UTXO of 1.0Mi. The value in this map for this `output.address` key will be -0.1Mi.
You probably took this into account, but I find it very confusing and error-prone, especially because on line 139 and 141 you could end up with a negative number!

>         addressesToValidate = SetUnion(dustAddressesToNumOfDustOutputs.keys, dustAddressesToNumOfDustOutputs.keys)

Reviewer 2: These two sets are the same?

The first option can cause issues, when dust outputs were burned before users could consolidate them. Also, changing the supply can be controversial.

The second option is much more complicated as it introduces a completely new unlock mechanisms and requires the nodes to store the Merkle tree roots indefinitely.

# Unresolved questions

- An attacker can send microtransactions to an address with a `SigLockedDustAllowanceOutput` in order to fill the allowed threshold and block honest senders of microtransactions. The owner of the address can mitigate this by simply consolidating the attacker's dust and collecting it for profit. The problem is that the cost of doing PoW too often may exceed the profit made by collecting dust. Perhaps it is a good idea to limit "tiny" (1 i) dust outputs more than larger dust outputs? In the original discussion each dust class was supposed to be capped to a certain fixed amount. But perhaps we should give weights to different dust classes. So "tiny" dust will weigh more and fill the cap more quickly than larger dust?

> - Total cap per address so db key won't be polluted... should this be part of the RFC? This also depends on how key-value dbs will behave at the implementation level. RocksDb for example can configured to handle this well.

Author: Not part of the RFC...

We might need to paginate the db keys... we should test on implementations.

# Motivation

An attacker, or even honest users, can proliferate the UTXO set with outputs holding a tiny amount of iotas. This can cause the UTXO set to grow to a prohibitively large size. 

> Nodes may stall or crash due to the increasing amount of memory and computational resources needed.

Reviewer: ```suggestion
Nodes may freeze or crash due to the increasing amount of memory and computational resources needed.
```

### Validation

Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 100,000 and rounded down, i.e. 10 outputs for each 1 Mi deposited.

> There is also a cap of 100 dust outputs per address we shouldn't exceed. 

Reviewer: ```suggestion
However, regardless of S, the number of dust outputs must never exceed 100 per address.
```

  - consuming a `SigLockedDustAllowanceOutput` on address A **or**
  - creating a dust output with address A,

> is only semantically valid, if, after T is booked, the number of unspent dust outputs on A does not exceed the allowed threshold of min(S / 100,000, 100).

Reviewer: ```suggestion
is only semantically valid, if, after T is booked, the number of unspent dust outputs on A does not exceed the allowed threshold of min(S / 100000, 100).
```
or `100_000` would be an option, but the comma is confusing
